torchvision/ops/focal_loss.py

```diff
@@ -56,3 +56,81 @@ def sigmoid_focal_loss(
+
+    Args:
+        inputs (Tensor): A float tensor of arbitrary shape.
+                The predictions for each example. Softmax() is applied on this tensor
+                to convert the raw logits to class probabilities. Expected shape is
+                (N, C, *).
+        targets (Tensor): Must be a long tensor similar to the one expected by
+                PyTorch's CrossEntropyLoss.
+                https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html
+                The class dimension is expected to be absent, and each
+                element is the class value in the range [0, C).
+        alpha (float): Weighting factor in range (0,1) to balance
+                positive vs negative examples or -1 for ignore. Default: ``0.25``.
+        gamma (float): Exponent of the modulating factor (1 - p_t) to
+                balance easy vs hard examples. Default: ``2``.
+        reduction (string): ``'none'`` | ``'mean'`` | ``'sum'``
+                ``'none'``: No reduction will be applied to the output.
+                ``'mean'``: The output will be averaged.
+                ``'sum'``: The output will be summed.
+                ``'instance-sum-batch-mean'``: The output will be summed for each
+                        value in the batch, and then averaged across the entire
+                        batch. Default: ``'none'``.
+    Returns:
+        Loss tensor with the reduction option applied.
+    """
+    # Adapted from this version by Thomas V.
+    # https://discuss.pytorch.org/t/focal-loss-for-imbalanced-multi-class-classification-in-pytorch/61289/2
+    # Referenced from this github issue:
+    # https://github.com/pytorch/vision/issues/3250
+    if not torch.jit.is_scripting() and not torch.jit.is_tracing():
+        _log_api_usage_once(softmax_focal_loss)
+
+    assert targets.dtype == torch.long, f"Expected a long tensor for 'targets', but got {targets.dtype}"
+
+    logits = inputs
+    ce_loss = nn.functional.cross_entropy(logits, targets, reduction='none')
+    # Instead of computing inputs.softmax(dim=1), we use the exponentiated
+    # negative log of the cross entropy loss.
+    #
+    # Why does this work?
+    # Since this is a multi-class setting, only one class is active. The
+    # probability of that class is 1, and the rest are all 0.
+    #
+    # Cross Entropy Loss computes:
+    # pt = softmax(...)
+    # loss = -1.0 * log(pt)
+    #
+    # Hence, exp(-loss) == pt
+    #
+    # This trick works only if the targets is a long tensor. If it's a float
+    # tensor, then each each value is a probability, and we'd need to divide
+    # the result of cross entropy loss by the probability, and hence would
+    # need to compute the softmax manually anyway. We don't implement that
+    # here for brevity, but this code can be extended for such a use-case.
+    pt = torch.exp(-ce_loss)
+    focal_loss = alpha * ((1 - pt) ** gamma) * ce_loss
```

Comment:
I have my doubts that the `alpha` term used here is correct. In fact we want `alpha_t`.

Although the paper doesn't explicitly give the formula for `alpha_t`, it states in it's definition

> For notational convenience, we define $`α_t$` analogously to how we defined $`p_t$`

So given that

$`
\begin{align*}
p_t =
\begin{cases}
\alpha & \text{if } y = 1 \\
(1 - p) & \text{otherwise}
\end{cases}
\end{align*}`$

I believe $`\alpha_t` should be analogously interpreted as 

$`
\begin{align*}
\alpha_t =
\begin{cases}
\alpha & \text{if } y = 1 \\
(1 - \alpha) & \text{otherwise}
\end{cases}
\end{align*}`$

This seems to be the case in torchvision

https://github.com/pytorch/vision/blob/cc0f9d02e1e7549e9e8ef1081a0d3b5d72f1297d/torchvision/ops/focal_loss.py#L43-L45

I've seen another implementation in the wild pass through `alpha` to the `weights` of `nll_loss`: 

[AdeelH/pytorch-multi-class-focal-loss/focal_loss.py](https://github.com/AdeelH/pytorch-multi-class-focal-loss/blob/2477bf68ac61d2d2e5e044c63f70b156073c5ffd/focal_loss.py#L48-L49)

This strikes me as the correct approach since it allows one to weigh each class separately in a multi-class setting.

Reply:
I agree with your assessment above.

I have put up a revised PR at https://github.com/pytorch/vision/pull/7760 since I switched local branches.